Added linkchecker and fixed broken links

anshuman-agarwala opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md and developer guide https://git.k8s.io/community/contributors/devel/development.md
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug

/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Adds linter to check links. Fixes broken links. Removes some deprecated/unneeded files.
**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #12104

**Special notes for your reviewer**:
I have removed equinix metal docs since equinix is shuttering it's metal service.
I've also removed weave works docs since the service has shut down.
**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

VannTen commented on .github/workflows/link-checker.yml:
```diff
@@ -0,0 +1,18 @@
+  linkchecker:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
```
Please use hash-pinned action in workflows (see the other workflows).

anshuman-agarwala replied:
Done

VannTen commented on contrib/terraform/equinix/README.md:
This probably should be at the very least in another commit, or in another PR.

VannTen commented on CONTRIBUTING.md:
```diff
@@ -44,4 +44,4 @@ Vagrant with VirtualBox or libvirt driver helps you to quickly spin test cluster
 7. Submit a pull request.
 8. Work with the reviewers on their suggestions.
-9. Ensure to rebase to the HEAD of your target branch and squash un-necessary commits (<https://blog.carbonfive.com/always-squash-and-rebase-your-git-commits/>) before final merger of your contribution.
+9. Ensure to rebase to the HEAD of your target branch and squash un-necessary commits (<https://gitscripts.com/git-rebase-i-squash>) before final merger of your contribution.
```
This line could actually go away completely, squash should not be an automatic reflex.

anshuman-agarwala replied:
Removed the line.

VannTen commented on docs/CSI/vsphere-csi.md:
```diff
@@ -63,7 +63,7 @@ kind: Pod
   name: nginx
 spec:
-  containers:
+  s:
```
??

anshuman-agarwala replied:
This was a mistake, sorry about that.

VannTen commented on docs/external_storage_provisioners/cephfs_provisioner.md:
````diff
@@ -40,26 +40,26 @@ The following example uses `cephfs-provisioner-1` as the identity for the instan
 ```
 
-Alternatively, deploy it in kubernetes, see [deployment](deploy/README.md).
+Alternatively, deploy it in kubernetes, see [deployment](https://github.com/kubernetes-retired/external-storage/blob/master/ceph/cephfs/deploy/README.md).
````
I'm not sure why we're switching to external link here ?

anshuman-agarwala replied:
I looked for the file this is referring to but could not find it in the repo.

yankay replied:
I completely agree.

tico88612 replied:
I have a few ideas here.
1. I think these outdated applications should be removed, and I have the same view as @yankay on this. They should also be phased out by other PRs.
2. I want to phase out the Kubespray's built-in application installations (e.g., ArgoCD, ingress-nginx, cephfs, rbd provisioned), whose templates we should not maintain. Avoid ingress-nginx CVE-liked scenarios. Instead, it should be up to the user to decide what to install (Design and use `roles/helm-apps` to handle this situation), and we only provide documentation examples to guide them.

VannTen replied:
Oh I hadn't realized we had code about this has well. Let's do that in another PR them. I thought it was only the documentation left.

VannTen commented on docs/external_storage_provisioners/rbd_provisioner.md:
````diff
@@ -52,20 +52,20 @@ The following example uses `rbd-provisioner-1` as the identity for the instance
 ```
 
-Alternatively, deploy it in kubernetes, see [deployment](deploy/README.md).
+Alternatively, deploy it in kubernetes, see [deployment](https://github.com/kubernetes-retired/external-storage/blob/master/ceph/cephfs/deploy/README.md).
````
Same as above

hegerdes commented:
Just a small insight from personal experience. This link check job could get very flaky on GH hosted runners.  
GH runners are often used for scraping, specially in the age of AI. They all use a fixed, publicly known Azure IP range which is shared by all GH users.

I used a link checker for my personal blog but hat to turn of the check for external links because of the may false positives I got due to being blocked by WAFs.  
So maybe this job should be optional, focus on internal links or use dedicated runners which have better IP reputation.

VannTen commented:
Are you still working on this ? I'm not sure what the delete of terraform are for ?

anshuman-agarwala commented:
> Are you still working on this ? I'm not sure what the delete of terraform are for ?

The delete in the terraform is because the equinix metal service has been shuttered: https://deploy.equinix.com/blog/an-update-on-the-latest-changes-to-metal-functionality/
I have addressed all the review comments, I'm happy to make any other changes that are required.

VannTen commented:
Ok, but that's not exactly related to a link checker, so that shouldn't go in this PR, (but welcome in another one)

anshuman-agarwala commented:
> Ok, but that's not exactly related to a link checker, so that shouldn't go in this PR, (but welcome in another one)

Removed the terraform changes and created #12229

VannTen commented on contrib/terraform/openstack/README.md:
```diff
@@ -10,7 +10,7 @@ most modern installs of OpenStack that support the basic services.
 ### Known compatible public clouds
 
-- [Auro](https://auro.io/)
+- [Auro](https://www.cacloud.com/)
```
Just remove it, since the domain changed, no idea if the cloud behave similarly enough

anshuman-agarwala replied:
Removed Auro

VannTen commented on docs/CNI/cilium.md:
````diff
@@ -355,7 +355,7 @@ To validate that Hubble UI is properly configured, set up a port forwarding for
 ```
 
-and then open [http://localhost:12000/](http://localhost:12000/).
+and then open <http://localhost:12000/>.
````
What's the difference here exactly ?

anshuman-agarwala replied:
It was required because the link checker was flagging the localhost links as unreachable. With this format it's ignoring these links.

VannTen replied:
Isn't that handled by the ignore file you added ?

anshuman-agarwala replied:
You're right, I've reverted this change.

VannTen commented on docs/cloud_controllers/openstack.md:
```diff
@@ -4,7 +4,7 @@
 Kubespray has been tested on a number of OpenStack Public Clouds including (in alphabetical order):
 
-- [Auro](https://auro.io/)
+- [Auro](https://www.cacloud.com/)
```
Same as above (and oh, we have duplicate docs :man_shrugging: )

anshuman-agarwala replied:
Removed Auro

VannTen commented:
> Just a small insight from personal experience. This link check job could get very flaky on GH hosted runners. GH runners are often used for scraping, specially in the age of AI. They all use a fixed, publicly known Azure IP range which is shared by all GH users.
> 
> I used a link checker for my personal blog but hat to turn of the check for external links because of the many false positives I got due to being blocked by WAFs. So maybe this job should be optional, focus on internal links or use dedicated runners which have better IP reputation.


That's a good point. There is also the matter of approval because of https://github.com/kubernetes-sigs/prow/issues/194
-> this is harder on the contrib ex front.

anshuman-agarwala commented on README.md:
```diff
@@ -227,6 +222,6 @@ See also [Network checker](docs/advanced/netcheck.md).
 [![Build graphs](https://gitlab.com/kargo-ci/kubernetes-sigs-kubespray/badges/master/pipeline.svg)](https://gitlab.com/kargo-ci/kubernetes-sigs-kubespray/-/pipelines)
 
-CI/end-to-end tests sponsored by: [CNCF](https://cncf.io), [Equinix Metal](https://metal.equinix.com/), [OVHcloud](https://www.ovhcloud.com/), [ELASTX](https://elastx.se/).
+CI/end-to-end tests sponsored by: [CNCF](https://cncf.io), [Equinix Metal](https://deploy.equinix.com/), [OVHcloud](https://www.ovhcloud.com/), [ELASTX](https://elastx.se/).
```
Removed Equinix and OVHcloud

VannTen reviewed (commented):
You need to rebase on top of current master to clear the CI stuff.
